media_cdn/dualtoken.rb

```diff
@@ -0,0 +1,179 @@
+# limitations under the License.
+
+# [START mediacdn_dualtoken_sign_token]
+require "base64"
+require "time"
+require "openssl"
+require "ed25519"
+
+class Dualtoken
+  def base64_encoder value
+    #   Returns a base64-encoded string compatible with Media CDN.
+    #
+    #   Media CDN uses URL-safe base64 encoding and strips off the padding at the end.
+
+    encoded_str = Base64.urlsafe_encode64(value).encode("utf-8")
+    encoded_str.delete "="
+  end
+
+  def sign_token(
+    base64_key:,
+    signature_algorithm:,
+    start_time: nil,
+    expiration_time: nil,
+    full_path: nil,
+    path_globs: nil,
+    url_prefix: nil,
+    session_id: nil,
+    data: nil,
+    headers: nil,
+    ip_ranges: nil
+  )
+
+    # Gets the signed URL suffix string for the Media CDN short token URL requests.
+    # One of (`url_prefix`, `full_path`, `path_globs`) must be included in each input.
+    # Args:
+    #     base64_key: a secret key as a base64 encoded string.
+    #     signature_algorithm: an algorithm as `SHA1`, `SHA256`, or `Ed25519`.
+    #     start_time: the start time as a UTC datetime object.
+    #     expiration_time: the expiration time as a UTC datetime object. If a value is not specified, the expiration time will be set to 5 mins from now.
+    #     url_prefix: the URL prefix to sign, including protocol.
+    #                 For example: http://example.com/path/ for URLs under /path or http://example.com/path?param=1
+    #     full_path:  a full path to sign, starting with the first '/'.
+    #                 For example: /path/to/content.mp4
+    #     path_globs: a set of ','- or '!'-delimited path glob strings.
+    #                 For example: /tv/*!/film/* to sign paths starting with /tv/ or /film/ in any URL.
+    #     session_id: a unique identifier for the session
+    #     data: an arbitrary data payload to include in the token
+    #     headers: the header name and value to be included in the signed token should follow name=value format.  May be specified more than once.
+    #                 For example: [{'name': 'foo', 'value': 'bar'}, {'name': 'baz', 'value': 'qux'}]
+    #     ip_ranges: a list of comma-separated IP ranges. Both IPv4 and IPv6 ranges are acceptable.
+    #                 For example: '203.0.113.0/24,2001:db8:4a7f:a732/64'
+    # Returns:
+    #     The Signed URL appended with the query parameters based on the
+    #     specified URL prefix and configuration.
+
+    decoded_key = Base64.urlsafe_decode64 base64_key
+    algo = signature_algorithm.downcase
+
+    # For most fields, the value we put in the token and the value we must sign
+    # are the same.  The FullPath and Headers use a different string for the
+    # value to be signed compared to the token.  To illustrate this difference,
+    # we'll keep the token and the value to be signed separate.
+    tokens = []
+    to_sign = []
+
+    # check for `full_path` or `path_globs` or `url_prefix`
+    if full_path
```

Comment:
`if full_path.nil?`
I think it's safer to specifically check for nils. Same for the other parameters.